frontend/src/metabase/visualizations/lib/apply_tooltips.js

```diff
@@ -6,160 +6,207 @@ import d3 from "d3";
-                     const card = series[seriesIndex].card;
-                     const isSingleSeriesBar = this.classList.contains("bar") && series.length === 1;
-                     const isArea = this.classList.contains("area");
-
-                     let data = [];
-                     if (Array.isArray(d.key)) { // scatter
-                         if (d.key._origin) {
-                             data = d.key._origin.row.map((value, index) => {
-                                 const col = d.key._origin.cols[index];
-                                 return { key: getFriendlyName(col), value: value, col };
-                             });
-                         } else {
-                             data = d.key.map((value, index) => (
-                                 { key: getFriendlyName(cols[index]), value: value, col: cols[index] }
-                             ));
-                         }
-                     } else if (d.data) { // line, area, bar
-                         if (!isSingleSeriesBar) {
-                             cols = series[seriesIndex].data.cols;
-                         }
-
-                         data = [
-                             {
-                                 key: getFriendlyName(cols[0]),
-                                 value: d.data.key,
-                                 col: cols[0]
-                             },
-                             {
-                                 key: getFriendlyName(cols[1]),
-                                 value: isNormalized
-                                      ? `${formatValue(d.data.value) * 100}%`
-                                      : d.data.value,
-                                 col: cols[1]
-                             }
-                         ];
-                     }
-
-                     if (data && series.length > 1) {
-                         if (card._breakoutColumn) {
-                             data.unshift({
-                                 key: getFriendlyName(card._breakoutColumn),
-                                 value: card._breakoutValue,
-                                 col: card._breakoutColumn
-                             });
-                         }
-                     }
-
-                     data = _.uniq(data, (d) => d.col);
-
-                     onHoverChange({
-                         // for single series bar charts, fade the series and highlght the hovered element with CSS
-                         index: isSingleSeriesBar ? -1 : seriesIndex,
-                         // for area charts, use the mouse location rather than the DOM element
-                         element: isArea ? null : this,
-                         event: isArea ? d3.event : null,
-                         data: data.length > 0 ? data : null,
+function setHoverFunctions(chart, series, isStacked, isNormalized, onHoverChange) {
```

Comment:
I feel like for some of these charting lib functions we should start adding comments for what each param actually contains and is used for (aside from / in addition to flow types). Maybe it's easier once you've been in it a bit but for example series doesn't immediately jump out as an array that might contain multiple series.

Reply:
Yeah that one got me for a while but in the charting lib `series` = an array of series. 😒 

I think `serieses` would ultimately be a better, more accurate name. Either way you're right these things could use some comments just so the things I figured out from poking around the code base don't need to be figured out again the hard way by the next unfortunate soul to visit here